File: scripts/dmzMBRAPluginDestroyObject.lua
local NodeType = dmz.object_type.new ("na_node")
local NodeLinkType = dmz.object_type.new ("na_link_attribute")

function receive (self, type, data)
   if dmz.data.is_a (data) then
      local handle = data:lookup_handle ("object", 1)
      if handle then 
         local link = self.objects[handle]
         if link then handle = link end
         local objType = dmz.object.type (handle)
         if
               objType and
               objType:is_of_type (NodeType) and
               dmz.object.is_object (handle) then
            local undoHandle = dmz.undo.start_record ("Delete Node")
            dmz.object.destroy (handle)
            dmz.undo.stop_record (undoHandle)
         elseif dmz.object.is_link (handle) then
            local undoHandle = dmz.undo.start_record ("Unlink Nodes")
            dmz.object.unlink (handle)
            dmz.undo.stop_record (undoHandle)
         end
      else self.log:error ("Got null handle!")
      end
   end
end

local DefaultCallbacks = {

destroy_object = function (self, object)
   self.objects[object] = nil
end,

}

local LinkCallbacks = {

update_link_object = function (self, link, attr, super, sub, object, prev)
   if object and dmz.object.type (object):is_of_type (NodeLinkType) then
      self.objects[object] = link
   end
end,

}

function new (config, name)

   local self = {
      log = dmz.log.new ("lua." .. name),
      message = config:to_message ("message.name", "DestroyObjectMessage"),
      obs = dmz.message_observer.new (name),
      objObs = dmz.object_observer.new (name),
      objects = {},
   }

   self.log:info ("Creating plugin: " .. name)
   self.log:info ("Message type: " .. self.message:get_name ())
   self.obs:register (self.message, receive, self)

   self.objObs:register (nil, DefaultCallbacks, self)
   self.objObs:register ("Node_Link", LinkCallbacks, self)

   return self
end

File: scripts/dmzMBRAPluginDestroyObject.js
var dmz =
   { object: require("dmz/components/object")
   , objectType: require("dmz/runtime/objectType")
   , data: require("dmz/runtime/data")
   , defs: require("dmz/runtime/definitions")
   , message: require("dmz/runtime/messaging")
   , undo: require("dmz/runtime/undo")
   }
   , NodeType = dmz.objectType.lookup ("na_node")
   , NodeLinkType = dmz.defs.createNamedHandle("Node_Link")
   , message = dmz.message.create (
         self.config.string("message.name", "DestroyObjectMessage"))
   , objects = []
   ;

message.subscribe (self, function (data) {
   if (dmz.data.isTypeOf (data)) {
      var handle = data.handle ("object", 0);
      if (handle) {
         var link = objects[handle];
         if (link) { handle = link; }
         var objType = dmz.object.type (handle);
         if (objType && objType.isOfType (NodeType) && dmz.object.isObject (handle)) {
            var undoHandle = dmz.undo.startRecord ("Delete Node");
            dmz.object.destroy (handle);
            dmz.undo.stopRecord (undoHandle);
         } else if (dmz.object.isLink (handle)) {
            var undoHandle = dmz.undo.startRecord ("Unlink Nodes");
            dmz.object.unlink (handle);
            dmz.undo.stopRecord (undoHandle);
         }
      } else { self.log.error ("Got null handle!"); }
   }
});

dmz.object.destroy.observe (self, function (object) { delete objects[object]; });

dmz.object.linkAttributeObject.observe (self, NodeLinkType,
function (link, attr, Super, sub, object, prev) {
   if (object && dmz.object.type (object).isOfType (NodeLinkType)) {
      objects[object] = link;
   }
});
